### PR TITLE
Minting with external consolidation (no attestation)

### PR DESCRIPTION
**Pull request by iamsahu**
## Description

This pull request introduces a new "consolidation" mechanism to the River contract, enabling the minting of LsETH specifically for ETH consolidation purposes. It adds new state, interface methods, events, and comprehensive tests to support and verify this functionality. The main changes are grouped below:

**Core Protocol Enhancements:**

* Added a new `BalanceToConsolidate` storage library (`BalanceToConsolidate.sol`) to track the amount of ETH pending consolidation, using unstructured storage.
* Introduced the `getBalanceToConsolidate()` view and `mintLsETHForConsolidation(uint256 _amount, address _recipient)` methods in the `RiverV1` contract, restricted to the keeper, to allow minting LsETH for consolidation and updating the consolidation buffer.

**Interface and Event Additions:**

* Extended the `IRiverV1` interface with the new consolidation methods and added two new events: `SetBalanceToConsolidate` (for tracking changes to the consolidation buffer) and `LsETHMintedForConsolidation` (for tracking minting actions). [[1]](diffhunk://#diff-b60cd97fe270d3777949b4d10e8f36471d394d54ea1275f115684e056a639063R101-R111) [[2]](diffhunk://#diff-b60cd97fe270d3777949b4d10e8f36471d394d54ea1275f115684e056a639063R254-R262)

**Testing Improvements:**

* Added a new test contract `RiverV1ConsolidationMintTests` to verify the correct behavior of consolidation minting, including access control, edge cases, and state updates.
* Updated test imports to include the new interface dependencies required for consolidation testing.

**Internal Refactoring:**

* Updated imports in `River.1.sol` to include the new `BalanceToConsolidate` state library.

## Notice

- [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/liquid-collective/liquid-collective-protocol/pulls) for the same update/change?
- [ ] Have you assigned this PR to yourself?
- [ ] Have you added at least 1 reviewer?
- [ ] Have you updated the official documentation?
- [ ] Have you added sufficient documentation in your code?
- [ ] Have you added relevant tests to the official test suite?

## Pull Request Type

- [ ] 💫 New Feature (Breaking Change)
- [ ] 💫 New Feature (Non-breaking Change)
- [ ] 🛠️ Bug fix (Non-breaking Change: Fixes an issue)
- [ ] 🕹️ Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)

## Testing

- [ ] Have you tested this code with the official test suite?
- [ ] Have you tested this code manually?

## Summary by CodeRabbit

* **New Features**
  * Added functions to query and track consolidation buffer balances.
  * Keeper can now mint LsETH with new consolidation events for balance updates.

* **Tests**
  * Added comprehensive test suite covering consolidation minting scenarios.
  * Updated formal verification specifications to account for consolidation balance in accounting rules.

### CODE REVIEW

File: contracts/src/interfaces/IRiver.1.sol [hunk 1]
```diff
@@ -99,6 +99,17 @@ interface IRiverV1 is IConsensusLayerDepositManagerV1, IUserDepositManagerV1, IS
...
+    /// @notice Emitted when the balance to consolidate is updated
+    /// @param oldAmount The old balance to consolidate
+    /// @param newAmount The new balance to consolidate
+    event SetBalanceToConsolidate(uint256 oldAmount, uint256 newAmount);
+
+    /// @notice Emitted when LsETH is minted for consolidation
+    /// @param recipient The address that received the minted LsETH
+    /// @param amountEth The amount of ETH attributed to consolidation
+    /// @param sharesMinted The amount of LsETH shares minted
+    event LsETHMintedForConsolidation(address indexed recipient, uint256 amountEth, uint256 sharesMinted);
```
Comment (juliaaschmidt) [inline contracts/src/interfaces/IRiver.1.sol:106]: amount ETH in (Wei)?

File: contracts/src/interfaces/IRiver.1.sol [hunk 2]
```diff
@@ -220,6 +231,15 @@ interface IRiverV1 is IConsensusLayerDepositManagerV1, IUserDepositManagerV1, IS
     /// @return True if slashing containment mode is active
     function getSlashingContainmentMode() external view returns (bool);
 
+    /// @notice Retrieve the current balance to consolidate
+    /// @return The current balance to consolidate
+    function getBalanceToConsolidate() external view returns (uint256);
+
+    /// @notice Mints LsETH to a recipient for consolidated ETH (keeper only)
```
Comment (juliaaschmidt) [inline contracts/src/interfaces/IRiver.1.sol:233]: (consolidator only)

File: contracts/src/River.1.sol [hunk 3]
```diff
@@ -174,6 +174,26 @@ contract RiverV1 is
...
     }
 
+    /// @inheritdoc IRiverV1
+    function getBalanceToConsolidate() external view returns (uint256) {
+        return ConsolidationBuffer.get();
+    }
+
+    /// @inheritdoc IRiverV1
+    function mintLsETHForConsolidation(uint256 _amount, address _recipient) external {
+        if (msg.sender != KeeperAddress.get()) {
```
Comment (juliaaschmidt) [inline contracts/src/River.1.sol:184]: consolidator?

File: contracts/src/River.1.sol [hunk 4]
```diff
@@ -174,6 +174,26 @@ contract RiverV1 is
         return BalanceToRedeem.get();
     }
 
+    /// @inheritdoc IRiverV1
+    function getBalanceToConsolidate() external view returns (uint256) {
+        return ConsolidationBuffer.get();
+    }
+
+    /// @inheritdoc IRiverV1
+    function mintLsETHForConsolidation(uint256 _amount, address _recipient) external {
```
Comment (juliaaschmidt) [inline contracts/src/River.1.sol:183]: Just a note: If I remember correctly the recipient and amount here are placeholders for what will be inside the attestation. We don't want to allow passing it in directly, it'll be the recipient will not be an arbitrary address but the withdrawal credentials of the attested consolidation.

File: contracts/src/River.1.sol [hunk 5]
```diff
@@ -174,6 +174,26 @@ contract RiverV1 is
...
+        if (msg.sender != KeeperAddress.get()) {
+            revert IConsensusLayerDepositManagerV1.OnlyKeeper();
+        }
+        if (_amount == 0) {
+            revert LibErrors.InvalidArgument();
+        }
+        LibSanitize._notZeroAddress(_recipient);
+        uint256 oldConsolidationBuffer = ConsolidationBuffer.get();
+        _setConsolidationBuffer(oldConsolidationBuffer, oldConsolidationBuffer + _amount);
+        uint256 sharesMinted = _mintShares(_recipient, _amount);
```
Comment (juliaaschmidt) [inline contracts/src/River.1.sol:193]: Could we check before this step that the recipient has the consolidate mask?

File: contracts/src/River.1.sol [hunk 6]
```diff
@@ -174,6 +174,26 @@ contract RiverV1 is
...
+    function mintLsETHForConsolidation(uint256 _amount, address _recipient) external {
+        if (msg.sender != KeeperAddress.get()) {
+            revert IConsensusLayerDepositManagerV1.OnlyKeeper();
+        }
+        if (_amount == 0) {
+            revert LibErrors.InvalidArgument();
+        }
+        LibSanitize._notZeroAddress(_recipient);
+        uint256 oldConsolidationBuffer = ConsolidationBuffer.get();
+        _setConsolidationBuffer(oldConsolidationBuffer, oldConsolidationBuffer + _amount);
```
Comment (juliaaschmidt) [inline contracts/src/River.1.sol:192]: you do not emit the `IRiver.SetBalanceToConsolidate event` as also flagged per AI review

File: contracts/src/interfaces/IRiver.1.sol (hunk 1 above)
Comment (iamsahu) [reply contracts/src/interfaces/IRiver.1.sol:106]: Fixed

File: contracts/src/River.1.sol (hunk 4 above)
Comment (iamsahu) [reply contracts/src/River.1.sol:183]: That's right.

File: contracts/src/River.1.sol (hunk 6 above)
Comment (iamsahu) [reply contracts/src/River.1.sol:192]: The event `SetConsolidationBuffer` is emitted in `_setConsolidationBuffer`. I have removed `SetBalanceToConsolidate` now.

File: contracts/src/River.1.sol (hunk 5 above)
Comment (iamsahu) [reply contracts/src/River.1.sol:193]: I have fixed this in 507

**Review by juliaaschmidt — COMMENTED** [review]
Great stuff! 🏆 👏 

Two questions:

1. We should introduce a new role here, that is not keeper no? Otherwise we have all roles centralised to keeper? cc @mischat 
2. Should we restrict recipients to EOA to minimise the risk of stranded inaccessible tokens? I guess no because we also want integrations with other contracts right?

**Review by juliaaschmidt — DISMISSED** [review]

File: contracts/src/interfaces/IRiver.1.sol (hunk 2 above)
Comment (iamsahu) [reply contracts/src/interfaces/IRiver.1.sol:233]: Will be addressed in another PR

File: contracts/src/River.1.sol (hunk 3 above)
Comment (iamsahu) [reply contracts/src/River.1.sol:184]: Will be addressed in another PR

**Review by juliaaschmidt — APPROVED** [review]
created this issue for the consolidator role #551

**Review by mischat — APPROVED** [review]
🛳️
